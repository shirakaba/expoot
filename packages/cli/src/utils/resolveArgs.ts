import { parseArgs } from 'node:util';

/**
 * Enables the resolution of arguments that can either be a string or a boolean.
 *
 * @param args arguments that were passed to the command.
 * @param rawMap raw map of arguments that are passed to the command.
 * @param extraArgs extra arguments and aliases that should be resolved as string or boolean.
 * @returns parsed arguments and project root.
 */
export async function resolveStringOrBooleanArgs(params: string[]) {
  const args = parseArgs({
    args: params,
    options: {
      help: {
        type: 'boolean',
        short: 'h',
      },
    },
    allowPositionals: true,
  });

  args = splitArgs(args);

  // Assert any missing arguments
  assertUnknownArgs(
    {
      ...rawMap,
      ...extraArgs,
    },
    args
  );

  // Collapse aliases into fully qualified arguments.
  args = collapseAliases(extraArgs, args);

  // Resolve all of the string or boolean arguments and the project root.
  return _resolveStringOrBooleanArgs({ ...rawMap, ...extraArgs }, args);
}
