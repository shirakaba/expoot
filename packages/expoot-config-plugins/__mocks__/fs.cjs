// https://vitest.dev/guide/mocking.html#example-3
const { fs } = require('memfs');

console.log('mocking fs with memfs');

module.exports = new Proxy(fs, {
  get(_target, prop, _receiver) {
    console.trace(`mock get "${prop.toString()}"`);
    return Reflect.get(...arguments);
  },
});

// module.exports = {
//   ...fs,
//   readdirSync: (path, options) => {
//     console.log('mock readdirSync');
//     fs.readdirSync(path, options);
//   },
// };
