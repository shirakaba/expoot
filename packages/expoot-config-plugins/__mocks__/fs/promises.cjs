// https://vitest.dev/guide/mocking.html#example-3

const { fs } = require('memfs');

console.log('mocking fs/promises with memfs');

module.exports = fs.promises;
