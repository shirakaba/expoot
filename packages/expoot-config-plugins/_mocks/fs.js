// From https://github.com/vitest-dev/vitest/discussions/5589#discussioncomment-9195492
import { vi } from 'vitest';

import { fs as memFs } from 'memfs';

vi.mock('fs', () => ({ ...memFs, default: memFs }));
vi.mock('fs/promises', () => memFs.promises);
vi.mock('node:fs', () => ({ ...memFs, default: memFs }));
vi.mock('node:fs/promises', () => ({ promises: memFs.promises }));
