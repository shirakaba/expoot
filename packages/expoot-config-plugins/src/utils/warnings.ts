import chalk from 'chalk';

import type { ModPlatform } from '../Plugin.types';

/**
 * Log a warning that doesn't disrupt the spinners.
 *
 * ```sh
 * » android: android.package: property is invalid https://expo.fyi/android-package
 * ```
 *
 * @param property Name of the config property that triggered the warning (best-effort)
 * @param text Main warning message
 * @param link Useful link to resources related to the warning
 */
export function addWarningAndroid(
  property: string,
  text: string,
  link?: string
) {
  addWarningForPlatform('android', property, text, link);
}

/**
 * Log a warning that doesn't disrupt the spinners.
 *
 * ```sh
 * » ios: ios.bundleIdentifier: property is invalid https://expo.fyi/bundle-identifier
 * ```
 *
 * @param property Name of the config property that triggered the warning (best-effort)
 * @param text Main warning message
 * @param link Useful link to resources related to the warning
 */
export function addWarningIOS(property: string, text: string, link?: string) {
  addWarningForPlatform('ios', property, text, link);
}

/**
 * Log a warning that doesn't disrupt the spinners.
 *
 * ```sh
 * » macos: macos.bundleIdentifier: property is invalid https://expo.fyi/bundle-identifier
 * ```
 *
 * @param property Name of the config property that triggered the warning (best-effort)
 * @param text Main warning message
 * @param link Useful link to resources related to the warning
 */
export function addWarningMacOS(property: string, text: string, link?: string) {
  addWarningForPlatform('macos', property, text, link);
}

/**
 * Log a warning that doesn't disrupt the spinners.
 *
 * ```sh
 * » windows: windows.guid: property is invalid
 * ```
 *
 * @param property Name of the config property that triggered the warning (best-effort)
 * @param text Main warning message
 * @param link Useful link to resources related to the warning
 */
export function addWarningWindows(
  property: string,
  text: string,
  link?: string
) {
  addWarningForPlatform('windows', property, text, link);
}

export function addWarningForPlatform(
  platform: ModPlatform & string,
  property: string,
  text: string,
  link?: string
) {
  console.warn(formatWarning(platform, property, text, link));
}

function formatWarning(
  platform: string,
  property: string,
  warning: string,
  link?: string
) {
  return chalk.yellow`${'» ' + chalk.bold(platform)}: ${property}: ${warning}${
    link ? chalk.gray(' ' + link) : ''
  }`;
}
