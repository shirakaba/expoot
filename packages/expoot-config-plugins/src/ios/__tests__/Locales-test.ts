import '../../../_mocks/fs.js';

import * as fs from 'node:fs';
import type FS from 'node:fs';

import * as WarningAggregator from '@expo/config-plugins/build/utils/warnings';
import { vol } from 'memfs';

import { readAllFiles } from '../../plugins/__tests__/fixtures/react-native-project';

import { getLocales, setLocalesAsync } from '../Locales';
import { getPbxproj } from '../utils/Xcodeproj';

import { getDirFromFS } from './utils/getDirFromFS';

vi.mock('@expo/config-plugins/build/utils/warnings');

const fsActual: typeof FS = await vi.importActual('node:fs');
const rnFixture = readAllFiles(fsActual);

describe('iOS Locales', () => {
  it('returns null if no values are provided', () => {
    expect(getLocales({})).toBeNull();
  });

  it('returns the locales object', () => {
    expect(
      getLocales({
        locales: {},
      })
    ).toStrictEqual({});
  });
});

describe('e2e: iOS locales', () => {
  const projectRoot = '/app';
  const platform = 'ios';
  beforeAll(async () => {
    vol.fromJSON(
      {
        'ios/testproject.xcodeproj/project.pbxproj':
          rnFixture['ios/HelloWorld.xcodeproj/project.pbxproj'],
        'ios/testproject/AppDelegate.m': '',
        'lang/fr.json': JSON.stringify({
          CFBundleDisplayName: 'french-name',
        }),
      },
      projectRoot
    );
  });

  afterAll(() => {
    vol.reset();
  });

  // Skip this because in `node_modules/xcode/lib/pbxProject.js`, the
  // `pbxProject.prototype.allUuids` function calls `for (key in sections) {`.
  // As the for..in loop does not declare `key` with `const` or `let`, the
  // runtime fails to parse the code.
  it.skip('writes all the image files expected', async () => {
    let project = getPbxproj(projectRoot, platform);

    project = await setLocalesAsync(
      {
        locales: {
          fr: 'lang/fr.json',
          // doesn't exist
          xx: 'lang/xx.json',
          // partially support inlining the JSON so our Expo Config type doesn't conflict with the resolved manifest type.
          es: { CFBundleDisplayName: 'spanish-name' },
        },
      },
      { project, projectRoot, platform }
    );
    // Sync the Xcode project with the changes.
    fs.writeFileSync(project.filepath, project.writeSync());

    const after = getDirFromFS(vol.toJSON(), projectRoot);
    const locales = Object.keys(after).filter((value) =>
      value.endsWith('InfoPlist.strings')
    );
    console.log('!! locales', locales);

    expect(locales.length).toBe(2);
    expect(after[locales[0]]).toMatchSnapshot();
    // Test that the inlined locale is resolved.
    expect(after[locales[1]]).toMatch(/spanish-name/);
    // Test a warning is thrown for an invalid locale JSON file.
    expect(WarningAggregator.addWarningIOS).toHaveBeenCalledWith(
      'locales.xx',
      'Failed to parse JSON of locale file for language: xx',
      'https://docs.expo.dev/distribution/app-stores/#localizing-your-ios-app'
    );
  });
});
