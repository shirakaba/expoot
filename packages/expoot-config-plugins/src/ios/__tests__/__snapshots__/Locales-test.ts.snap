// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e: iOS locales > writes all the image files expected 1`] = `"CFBundleDisplayName = "french-name";"`;

exports[`e2e: iOS locales writes all the image files expected 1`] = `"CFBundleDisplayName = "french-name";"`;
